### PR TITLE
fix: correct total_num_steps and batch_size calculation with context parallelism

### LINKED ISSUE #3218: Wrong epoch when turning on `context_parallel_size `
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

The number of epochs should remain consistent regardless of whether `context_parallel_size ` is configured or not.

### Current behaviour

When it **is set**, the training `final` log behaves as follows:
```
{'train_runtime': 79.4145, 'train_samples_per_second': 25.789, 'train_steps_per_second': 0.201, 'train_loss': 3.3714769408106804, 'memory/max_mem_active(gib)': 13.62, 'memory/max_mem_allocated(gib)': 13.62, 'memory/device_mem_reserved(gib)': 14.68, 'epoch': 5.42}
```

When it is not set, the training `final` log behave as follows:
```
{'train_runtime': 49.8935, 'train_samples_per_second': 5.131, 'train_steps_per_second': 0.04, 'train_loss': 5.478998422622681, 'memory/max_mem_active(gib)': 46.82, 'memory/max_mem_allocated(gib)': 46.82, 'memory/device_mem_reserved(gib)': 47.9, 'epoch': 0.84}
```

### Steps to reproduce

axolotl train placeholder.yaml

### Config yaml

```yaml
base_model: /model/Llama-3.2-1B-Instruct

plugins:
  - axolotl.integrations.liger.LigerPlugin
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_layer_norm: true
liger_fused_linear_cross_entropy: true
# processor_type: AutoProcessor
remove_unused_columns: false

shuffle_merged_datasets: true

# Training configuration
output_dir: /data/axolotl/outputs/Llama3.2-1B-test-1015-wo-cp
num_epochs: 1

# Sequence and packing settings
sequence_len: 32768
sample_packing: true
pad_to_sequence_len: true
pretrain_multipack_attn: true  # Prevent cross-attention between packed sequences
flash_attention: true
# context_parallel_size: 8   # You should change this to survey.
sample_packing_bin_size: 1000
sample_packing_group_size: 200000

# Batch size settings
gradient_accumulation_steps: 16
micro_batch_size: 1

# Optimizer and scheduler
optimizer: adamw_torch_fused
lr_scheduler: cosine
cosine_min_lr_ratio: 0.1
learning_rate: 2e-4
warmup_ratio: 0.05
weight_decay: 0.01

# Precision and performance
bf16: true
tf32: true

# Logging and checkpointing
logging_steps: 1
save_strategy: steps
save_steps: 2000
save_total_limit: 3

# Data
dataset_prepared_path: /data/axolotl/prepared_path/Llama3.2-1B-test-1015
dataloader_num_workers: 4
dataset_processes: 128
deepspeed: /data/axolotl/deepspeed_configs/zero1.json

special_tokens:
  pad_token: <|finetune_right_pad_id|>

datasets:
    - path: /data/datasets/pt/fineweb_sample10BT/fineweb_sample10BT.parquet
      ds_type: parquet
      type: completion
      field: text
      split: "train[:10000]"
```

### Possible solution

_No response_

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12.3

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by Yatimai**
<html>
<body>
<html><head></head><body><h1>Fix total_num_steps and batch_size calculation with context parallelism</h1>
<h1>Description</h1>
<p>Two bugs in <code>calculate_total_num_steps()</code> and <code>normalize_config()</code> caused incorrect epoch counting when using <code>context_parallel_size</code>.</p>
<p><strong>Bug 1 — <code>normalize_config()</code> in <code>config/__init__.py</code>:</strong>
<code>batch_size</code> was multiplied by <code>world_size</code>, but with context parallelism the effective data-parallel world size is <code>world_size // context_parallel_size</code>. With CP=2 on 2 GPUs, <code>batch_size</code> was 8 instead of 4, causing the model to process only 50% of the dataset per epoch.</p>
<p><strong>Bug 2 — <code>calculate_total_num_steps()</code> in <code>trainer.py</code>:</strong>
<code>total_num_steps</code> was multiplied by <code>context_parallel_size</code> at 3 locations. This was a compensatory workaround for Bug 1, but it inflated <code>max_steps</code>, causing the HF Trainer to report ~5.4 epochs instead of ~1.0 (as reported in #3218).</p>
<p><strong>Fix:</strong></p>
<ul>
<li><code>config/__init__.py</code>: compute <code>effective_world_size = world_size // (context_parallel_size or 1)</code> and use it for <code>batch_size</code></li>
<li><code>trainer.py</code>: remove the 3 <code>* cfg.context_parallel_size</code> multiplications in <code>calculate_total_num_steps()</code></li>
</ul>
<p>This is consistent with how HuggingFace Trainer computes <code>dp_world_size</code> internally (<code>world_size // tp_size // cp_size // sp_size</code>).</p>
<h2>Motivation and Context</h2>
<p>Fixes #3218</p>
<h2>How has this been tested?</h2>
<p>Tested on 2x H100 SXM (NVLink) with TinyLlama-1.1B, <code>mhenrichsen/alpaca_2k_test</code> dataset, DeepSpeed ZeRO-1, <code>num_epochs: 1</code>.</p>

Config | batch_size | total_num_steps | tokens processed | epoch
-- | -- | -- | -- | --
CP=2 (before fix) | 8 | 250 | 1,024,000 (50%) | 1.0
CP=2 (after fix) | 4 | 500 | 2,048,000 (100%) | 2.0*
No CP (reference) | 8 | 500 | 2,048,000 (100%) | 1.0

<p>*The reported epoch=2.0 with CP is a separate issue: <code>DistributedSampler</code> shards by <code>world_size</code> instead of <code>dp_world_size</code>, so the Trainer sees 250 steps per "dataloader epoch" and counts 500/250=2. This affects epoch display and LR schedule but not the actual data processed. Fixing the sampler would require a custom <code>DistributedSampler</code> for CP groups (similar to the pattern in <code>core/trainers/grpo/sampler.py</code>), which is a deeper change best addressed separately.</p>
<h2>AI Usage Disclaimer</h2>
<p>Yes — Claude was used for code analysis and test planning.</p>
<h2>Types of changes</h2>
<ul>
<li>[x] Bug fix (non-breaking change which fixes an issue)</li>
</ul></body></html>
</body>
</html>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed batch size scaling calculation in distributed training with context parallelism, now using an effective world size computation.
  * Corrected total training step calculations by removing incorrect context parallel size factors, affecting step estimation across different training configurations.

### CODE REVIEW

**Comment by NanoCode012** [comment]
Thanks the change looks good to me.

Following your point, I suspect we'll need to make a followup PR for TP as well (?)

> DistributedSampler shards by world_size instead of dp_world_size, so the Trainer sees 250 steps per "dataloader epoch" and counts 500/250=2. This affects epoch display and LR schedule but not the actual data processed.

Let me see if we can just monkeypatch the code, but wouldn't affecting the LR scheduler be a significant drawback too, or do you mean this is just a logging side-effect?

**Comment by Yatimai** [comment]
Thanks for the review.

TP: yes, I suspect the same pattern applies. Happy to look into it as a follow-up PR.

LR scheduler: the scheduler uses `total_num_steps` (now correct at 500), so the actual learning rate curve is unaffected. The only side-effect is the epoch counter in logs: `DistributedSampler` shards by `world_size` instead of `dp_world_size`, so the Trainer sees 250 steps per "dataloader epoch" and reports 2.0 instead of 1.0. The data processed and LR schedule are both correct. Fixing the sampler would need a custom `DistributedSampler` for CP groups, which I think is better as a separate PR.

**Review by NanoCode012 — APPROVED** [review]
TODOS in followup:
- [ ] Create custom Sampler like GRPO sampler to handle LR scheduler (not as simple as return DistributedSampler with num_replica in get_train_sampler Trainer)
- [ ] Fix TP calculation impacting batch size too

Just waiting CI before tag ready to merge
